### DISCUSSION TITLE
Enum flag patterns

### DISCUSSION THREAD

**Discussion by oriko1010** [dotnet/csharplang#4012] [Language Ideas] [answered]
I often find myself writing similar patterns with flag enums where certain flags take precedence
```csharp
direction switch
{
    _ when direction.HasFlag(Direction.Up) => ...,
    _ when direction.HasFlag(Direction.Down) => ...,
    Direction.Left => ...,
    Direction.Right => ...,
};
```
It would be really convenient to have a `flag constant-expression` pattern to match such cases (exact syntax doesn't matter).
```csharp
direction switch
{
    flag Direction.Up => ...,
    flag Direction.Down => ...,
    flag Direction.Left => ...,
    flag Direction.Right => ...,
};
```
Has there been any discussion on this? I wouldn't mind trying to write a proposal or maybe even attempt an implementation if this is something desirable.

**Comment by YairHalberstadt** [comment]
This does seem interesting, but I'm not sure how practically useful it is:

With flags, you generally want to handle all flags.
In a switch you can only handle one case.
So I'm not sure how often you would actually switch on flags.

In practice in your case for example I would imagine you want to handle all these cases separately:

```csharp
0
Up
Down
Left
Right
Up Left
Up Right
Down Left
Down Right
```
flags wouldn't be approriate for this, whilst a normal switch on enum values would.

**Comment by alrz** [comment, marked as answer]
With https://github.com/dotnet/csharplang/issues/1047 I could  image you'd be able to do something like that 
```cs
static bool Flag(this Direction self, Direction d) => self.HasFlags(d); // hypothetical "pattern extensions"

direction switch
{
 Flag(Direction.Up) => ...,
 Flag(Direction.Down) => ...,
}
```

**Reply by oriko1010** [reply]
Perfect, that's good enough.
